### PR TITLE
@gasket/plugin-config_error_handling

### PR DESCRIPTION
**Pull request by mmason2-godaddy**
## Summary
Descriptive error handling when hooking into the `appRequestConfig` lifecycles. Currently if a config object is not returned from the handlers it results in an error [here](https://github.com/godaddy/gasket/blob/main/packages/gasket-plugin-config/lib/middleware.js#L20).

Added descriptive error handling for `appEnvConfig` lifecycle [here](https://github.com/godaddy/gasket/blob/main/packages/gasket-plugin-config/lib/plugin.js#L12-L15)

## Changelog
Unclear if this change should be added to the change log.

## Test Plan
`appRequestConfig` - `middleware.test.js` - Added test case to ensure that an error is thrown and has a matching description.
`appEnvConfig` - `plugin.test.js` - Added test case to ensure that an error is thrown and has a matching description.

### CODE REVIEW

File: packages/gasket-plugin-config/lib/middleware.js [hunk 1]
```diff
@@ -16,6 +16,17 @@ function handler(gasket) {
...
         res);
 
+      // Error handling
+      // Better error description for the appRequestConfig & appEnvConfig lifecycles
+      // In the event a config object isn't returned from the hook(s)
+      // it will result in a error in the destructuring of req.config
+      if (!req.config) {
+        throw new Error(
+          // eslint-disable-next-line max-len
+          'req.config is undefined - Ensure that all plugins hooking into the appRequestConfig and appEnvConfig lifecycle return a config object.'
```
Comment (kinetifex) [inline packages/gasket-plugin-config/lib/middleware.js:26]: 👍🏻 Let's make the error message concise and specific. Can you do a check and throw for `appEnvConfig` [here](https://github.com/godaddy/gasket/blob/main/packages/gasket-plugin-config/lib/plugin.js#L12-L15).

You messaging could then be something like:
`An appRequestConfig lifecycle hook did not return a config object.`
and
`An appEnvConfig lifecycle hook did not return a config object.`

**Review by kinetifex — APPROVED** [review]
Great investigation and contribution! I left a couple of small cleaning comments - then 👍🏻

File: packages/gasket-plugin-config/test/plugin.test.js [hunk 2]
```diff
@@ -155,6 +155,23 @@ describe('Plugin', () => {
...
+      Object.assign(gasket.config, {
+        root: path.join(__dirname, './fixtures', 'templated'),
+        env: 'test'
+      });
+
+      gasket.execWaterfall.mockImplementation((event) => {
+        expect(event).toEqual('appEnvConfig');
+        // eslint-disable-next-line no-undefined
+        return undefined;
+      });
```
[the remark this answers is not in the document]
Comment (mmason2-godaddy) [reply packages/gasket-plugin-config/test/plugin.test.js:168]: Yeah this was definitely a brain lapse on my part 🤦‍♂️ 

Great catch!

File: packages/gasket-plugin-config/lib/middleware.js [hunk 3]
```diff
@@ -16,6 +16,16 @@ function handler(gasket) {
         req,
         res);
 
+      // Error handling
+      // Better error description for the appRequestConfig lifecycles
+      // In the event a config object isn't returned from the hook(s)
+      // it will result in a error in the destructuring of req.config
```
Comment (kawikabader) [inline packages/gasket-plugin-config/lib/middleware.js:22]: Is this comment necessary?

File: packages/gasket-plugin-config/lib/plugin.js [hunk 4]
```diff
@@ -13,6 +13,13 @@ module.exports = {
         'appEnvConfig',
         mergeRootConfig(gasket, mergeConfigFiles(gasket))
       );
+
+      // Better error description for the appRequestConfig lifecycles
```
Comment (kawikabader) [inline packages/gasket-plugin-config/lib/plugin.js:17]: Do we need this comment?

File: packages/gasket-plugin-config/lib/middleware.js [hunk 5]
```diff
@@ -16,6 +16,16 @@ function handler(gasket) {
...
+      if (!req.config) {
+        throw new Error(
```
Comment (kawikabader) [inline packages/gasket-plugin-config/lib/middleware.js:24]: Should we use the `gasket.logger` object here?

File: packages/gasket-plugin-config/lib/plugin.js [hunk 6]
```diff
@@ -13,6 +13,13 @@ module.exports = {
...
+      if (!gasket[ENV_CONFIG]) {
+        throw new Error(
```
Comment (kawikabader) [inline packages/gasket-plugin-config/lib/plugin.js:19]: Should we use the `gasket.logger` object here?

**Review by kawikabader — APPROVED** [review]
Just a couple of questions.

File: packages/gasket-plugin-config/lib/middleware.js (hunk 5 above)
Comment (kinetifex) [reply packages/gasket-plugin-config/lib/middleware.js:24]: I think a hard throw on these here makes sense. It should fail the app.

File: packages/gasket-plugin-config/lib/middleware.js (hunk 3 above)
Comment (mmason2-godaddy) [reply packages/gasket-plugin-config/lib/middleware.js:22]: It's verbose and could be removed.

**Pull request merged**